File: hausverwaltung/hausverwaltung/report/kautionskonten/kautionskonten.py
from __future__ import annotations

from collections import defaultdict
from datetime import date
from typing import Any

import frappe
from frappe import _
from frappe.utils import flt, getdate, today

from hausverwaltung.hausverwaltung.utils.report_helpers import enrich_link_titles


TOLERANCE = 0.01


def execute(filters=None):
	filters = frappe._dict(filters or {})
	_apply_defaults(filters)

	rows = _get_contract_rows(filters)
	if not rows:
		return get_columns(), [], None, None, _get_summary([], filters)

	amounts = _get_kaution_amounts([row.mietvertrag for row in rows], filters.stichtag)
	balances = _get_gl_balances(
		[row.gl_account for row in rows if row.get("gl_account")],
		filters.stichtag,
		filters.get("company"),
	)
	currencies = _get_company_currencies(rows, filters.get("company"))

	for row in rows:
		amount_info = amounts.get(row.mietvertrag) or {}
		row["kaution_betrag"] = amount_info.get("betrag")
		row["kaution_ab"] = amount_info.get("von")
		row["kaution_gepflegt"] = bool(amount_info)
		row["saldo"] = balances.get(row.gl_account, 0.0) if row.get("gl_account") else None
		row["currency"] = _row_currency(row, currencies, filters.get("company"))
		row["differenz"] = _difference(row)
		row["pruefung"] = _status(row)
		row["mietvertrag_name"] = row.get("kunde_anzeige") or row.get("kunde")

	rows = _sort_rows(rows)
	columns = get_columns()
	enrich_link_titles(rows, columns)
	return columns, rows, None, None, _get_summary(rows, filters)


def _apply_defaults(filters):
	filters.stichtag = getdate(filters.get("stichtag") or today())
	filters.nur_aktive_vertraege = int(filters.get("nur_aktive_vertraege", 1) or 0)
	filters.nur_mit_kautionskonto = int(filters.get("nur_mit_kautionskonto", 1) or 0)


def _get_contract_rows(filters) -> list[frappe._dict]:
	conditions = ["mv.docstatus < 2"]
	params: dict[str, Any] = {"stichtag": filters.stichtag}

	if filters.get("immobilie"):
		conditions.append("mv.immobilie = %(immobilie)s")
		params["immobilie"] = filters.immobilie

	if filters.nur_aktive_vertraege:
		conditions.append(
			"((mv.von IS NULL OR mv.von <= %(stichtag)s)"
			" AND (mv.bis IS NULL OR mv.bis >= %(stichtag)s))"
		)

	if filters.nur_mit_kautionskonto:
		conditions.append("COALESCE(mv.kautionskonto, '') != ''")

	if filters.get("company"):
		conditions.append(
			"("
			"COALESCE(ba.company, '') = %(company)s"
			" OR COALESCE(acc.company, '') = %(company)s"
			" OR COALESCE(mv.kautionskonto, '') = ''"
			")"
		)
		params["company"] = filters.company

	where = " AND ".join(conditions)
	return frappe.db.sql(
		f"""
		SELECT
			mv.name AS mietvertrag,
			mv.wohnung AS wohnung,
			mv.immobilie AS immobilie,
			mv.kunde AS kunde,
			c.customer_name AS kunde_anzeige,
			mv.status AS vertragsstatus,
			mv.von AS von,
			mv.bis AS bis,
			mv.kautionskonto AS kautionskonto,
			mv.kaution_notizen AS kaution_notizen,
			ba.account_name AS bank_account_name,
			ba.bank AS bank,
			ba.iban AS iban,
			ba.account AS gl_account,
			ba.company AS bank_account_company,
			acc.account_name AS gl_account_name,
			acc.account_number AS gl_account_number,
			acc.company AS gl_company
		FROM `tabMietvertrag` mv
		LEFT JOIN `tabCustomer` c ON c.name = mv.kunde
		LEFT JOIN `tabBank Account` ba ON ba.name = mv.kautionskonto
		LEFT JOIN `tabAccount` acc ON acc.name = ba.account
		WHERE {where}
		ORDER BY mv.immobilie, mv.wohnung, mv.von, mv.name
		""",
		params,
		as_dict=True,
	)


def _get_kaution_amounts(mietvertraege: list[str], stichtag: date) -> dict[str, dict[str, Any]]:
	if not mietvertraege:
		return {}

	rows = frappe.db.sql(
		"""
		SELECT parent AS mietvertrag, von, miete AS betrag, idx
		FROM `tabStaffelmiete`
		WHERE parenttype = 'Mietvertrag'
		  AND parentfield = 'kaution'
		  AND parent IN %(mietvertraege)s
		ORDER BY parent, von, idx
		""",
		{"mietvertraege": tuple(mietvertraege)},
		as_dict=True,
	)

	grouped: dict[str, list[frappe._dict]] = defaultdict(list)
	for row in rows:
		grouped[row.mietvertrag].append(row)

	result: dict[str, dict[str, Any]] = {}
	for mietvertrag, staffeln in grouped.items():
		selected = None
		for row in staffeln:
			if not row.get("von"):
				if selected is None:
					selected = row
				continue
			if getdate(row.von) <= stichtag:
				selected = row
		if selected:
			result[mietvertrag] = {
				"betrag": flt(selected.get("betrag")),
				"von": selected.get("von"),
			}
	return result


def _get_gl_balances(accounts: list[str], stichtag: date, company: str | None) -> dict[str, float]:
	accounts = sorted(set(accounts))
	if not accounts:
		return {}

	conditions = [
		"account IN %(accounts)s",
		"posting_date <= %(stichtag)s",
		"is_cancelled = 0",
	]
	params: dict[str, Any] = {"accounts": tuple(accounts), "stichtag": stichtag}
	if company:
		conditions.append("company = %(company)s")
		params["company"] = company

	rows = frappe.db.sql(
		f"""
		SELECT account, SUM(debit) AS debit, SUM(credit) AS credit
		FROM `tabGL Entry`
		WHERE {" AND ".join(conditions)}
		GROUP BY account
		""",
		params,
		as_dict=True,
	)
	return {row.account: flt(row.debit) - flt(row.credit) for row in rows}


def _get_company_currencies(rows: list[frappe._dict], selected_company: str | None) -> dict[str, str | None]:
	companies = {selected_company} if selected_company else set()
	for row in rows:
		if row.get("gl_company"):
			companies.add(row.gl_company)
		if row.get("bank_account_company"):
			companies.add(row.bank_account_company)
	companies.discard(None)
	companies.discard("")
	if not companies:
		return {}

	company_rows = frappe.get_all(
		"Company",
		filters={"name": ("in", sorted(companies))},
		fields=["name", "default_currency"],
	)
	return {row.name: row.default_currency for row in company_rows}


def _row_currency(row, currencies: dict[str, str | None], selected_company: str | None):
	if selected_company and currencies.get(selected_company):
		return currencies[selected_company]
	return currencies.get(row.get("gl_company")) or currencies.get(row.get("bank_account_company"))


def _difference(row) -> float | None:
	if not row.get("gl_account") or row.get("saldo") is None or not row.get("kaution_gepflegt"):
		return None
	return flt(row.get("saldo")) - flt(row.get("kaution_betrag"))


def _status(row) -> str:
	if not row.get("kautionskonto"):
		return _("Kautionskonto fehlt")
	if not row.get("kaution_gepflegt"):
		return _("Kaution fehlt")
	diff = flt(row.get("differenz"))
	if abs(diff) <= TOLERANCE:
		return _("OK")
	if diff < 0:
		return _("Unterdeckt")
	return _("Überdeckt")


def _sort_rows(rows: list[frappe._dict]) -> list[frappe._dict]:
	status_rank = {
		"Kautionskonto fehlt": 0,
		"Kaution fehlt": 1,
		"Unterdeckt": 2,
		"Überdeckt": 3,
		"OK": 4,
	}
	return sorted(
		rows,
		key=lambda row: (
			status_rank.get(row.get("pruefung"), 9),
			row.get("immobilie") or "",
			row.get("wohnung") or "",
			row.get("mietvertrag") or "",
		),
	)


def _get_summary(rows: list[frappe._dict], filters) -> list[dict[str, Any]]:
	currency = None
	for row in rows:
		if row.get("currency"):
			currency = row.currency
			break

	total_kaution = sum(flt(row.get("kaution_betrag")) for row in rows if row.get("kaution_gepflegt"))
	total_saldo = sum(flt(row.get("saldo")) for row in rows if row.get("saldo") is not None)
	diff = total_saldo - total_kaution
	problem_count = sum(1 for row in rows if row.get("pruefung") != "OK")

	return [
		{
			"value": len(rows),
			"indicator": "blue",
			"label": _("Kautionskonten"),
			"datatype": "Int",
		},
		{
			"value": problem_count,
			"indicator": "red" if problem_count else "green",
			"label": _("Auffällig"),
			"datatype": "Int",
		},
		{
			"value": total_kaution,
			"indicator": "blue",
			"label": _("Kaution laut Vertrag"),
			"datatype": "Currency",
			"currency": currency,
		},
		{
			"value": total_saldo,
			"indicator": "blue",
			"label": _("Saldo"),
			"datatype": "Currency",
			"currency": currency,
		},
		{
			"value": diff,
			"indicator": "green" if abs(diff) <= TOLERANCE else "orange",
			"label": _("Differenz"),
			"datatype": "Currency",
			"currency": currency,
		},
	]


def get_columns():
	return [
		{
			"label": _("Prüfung"),
			"fieldname": "pruefung",
			"fieldtype": "Data",
			"width": 115,
		},
		{
			"label": _("Mieter"),
			"fieldname": "mietvertrag",
			"fieldtype": "Link",
			"options": "Mietvertrag",
			"width": 230,
		},
		{
			"label": _("Immobilie"),
			"fieldname": "immobilie",
			"fieldtype": "Link",
			"options": "Immobilie",
			"width": 150,
		},
		{
			"label": _("Wohnung"),
			"fieldname": "wohnung",
			"fieldtype": "Link",
			"options": "Wohnung",
			"width": 150,
		},
		{
			"label": _("Von"),
			"fieldname": "von",
			"fieldtype": "Date",
			"width": 95,
		},
		{
			"label": _("Bis"),
			"fieldname": "bis",
			"fieldtype": "Date",
			"width": 95,
		},
		{
			"label": _("Kautionskonto"),
			"fieldname": "kautionskonto",
			"fieldtype": "Link",
			"options": "Bank Account",
			"width": 220,
		},
		{
			"label": _("IBAN"),
			"fieldname": "iban",
			"fieldtype": "Data",
			"width": 190,
		},
		{
			"label": _("Bank"),
			"fieldname": "bank",
			"fieldtype": "Link",
			"options": "Bank",
			"width": 130,
		},
		{
			"label": _("Sachkonto"),
			"fieldname": "gl_account",
			"fieldtype": "Link",
			"options": "Account",
			"width": 220,
		},
		{
			"label": _("Kaution ab"),
			"fieldname": "kaution_ab",
			"fieldtype": "Date",
			"width": 100,
		},
		{
			"label": _("Kaution laut Vertrag"),
			"fieldname": "kaution_betrag",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 145,
		},
		{
			"label": _("Saldo"),
			"fieldname": "saldo",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{
			"label": _("Differenz"),
			"fieldname": "differenz",
			"fieldtype": "Currency",
			"options": "currency",
			"width": 120,
		},
		{
			"label": _("Vertragsstatus"),
			"fieldname": "vertragsstatus",
			"fieldtype": "Data",
			"width": 110,
		},
		{
			"label": _("Notizen"),
			"fieldname": "kaution_notizen",
			"fieldtype": "Data",
			"width": 220,
		},
	]

File: hausverwaltung/hausverwaltung/report/kautionskonten/kautionskonten.js
frappe.query_reports["Kautionskonten"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Firma"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "stichtag",
			label: __("Stichtag"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "immobilie",
			label: __("Immobilie"),
			fieldtype: "Link",
			options: "Immobilie",
		},
		{
			fieldname: "nur_aktive_vertraege",
			label: __("Nur aktive Mietverträge"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "nur_mit_kautionskonto",
			label: __("Nur mit Kautionskonto"),
			fieldtype: "Check",
			default: 1,
		},
	],

	onload: function (report) {
		report.page.add_inner_button(__("Drucken"), () => {
			open_kautionskonten_compact_print(report);
		});
		report.page.add_inner_button(__("PDF"), () => {
			open_kautionskonten_print_dialog(report, true);
		});
	},

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "pruefung" && data?.pruefung) {
			const indicator = {
				OK: "green",
				Unterdeckt: "red",
				Überdeckt: "orange",
				"Kautionskonto fehlt": "red",
				"Kaution fehlt": "orange",
			}[data.pruefung] || "gray";
			return `<span class="indicator-pill ${indicator}">${__(data.pruefung)}</span>`;
		}
		if (column.fieldname === "differenz" && Math.abs(data?.differenz || 0) > 0.01) {
			return `<strong>${value || ""}</strong>`;
		}
		return value;
	},
};

const KAUTIONSKONTEN_PRINT_COLUMNS = [
	"pruefung",
	"mietvertrag",
	"immobilie",
	"wohnung",
	"kautionskonto",
	"iban",
	"kaution_betrag",
	"saldo",
	"differenz",
	"kaution_notizen",
];

const KAUTIONSKONTEN_COMPACT_COLUMNS = [
	{ fieldname: "pruefung", label: __("Prüfung"), type: "text" },
	{ fieldname: "mietvertrag", label: __("Mieter"), type: "link_label" },
	{ fieldname: "immobilie", label: __("Immobilie"), type: "text" },
	{ fieldname: "wohnung", label: __("Wohnung"), type: "text" },
	{ fieldname: "kautionskonto", label: __("Kautionskonto"), type: "text" },
	{ fieldname: "iban", label: __("IBAN"), type: "text" },
	{ fieldname: "kaution_betrag", label: __("Kaution"), type: "currency" },
	{ fieldname: "saldo", label: __("Saldo"), type: "currency" },
	{ fieldname: "differenz", label: __("Diff."), type: "currency" },
	{ fieldname: "kaution_notizen", label: __("Notizen"), type: "text" },
];

function open_kautionskonten_print_dialog(report, as_pdf) {
	const dialog = frappe.ui.get_print_settings(
		false,
		(print_settings) => {
			print_settings.orientation = "Landscape";
			print_settings.include_filters = 1;
			if (!print_settings.pick_columns) {
				print_settings.columns = KAUTIONSKONTEN_PRINT_COLUMNS.filter((fieldname) =>
					report.columns?.some((column) => column.fieldname === fieldname)
				);
			}
			if (as_pdf) {
				report.pdf_report(print_settings);
			} else {
				report.print_report(print_settings);
			}
		},
		report.report_doc?.letter_head,
		report.get_visible_columns()
	);
	report.add_portrait_warning?.(dialog);
}

function open_kautionskonten_compact_print(report) {
	const rows = report.data || [];
	if (!rows.length) {
		frappe.msgprint(__("Keine Daten zum Drucken vorhanden."));
		return;
	}

	const print_window = window.open("", "_blank");
	if (!print_window) {
		frappe.msgprint(__("Der Browser hat das Druckfenster blockiert."));
		return;
	}

	const filters = build_kautionskonten_filter_text(report);
	const generated_at = frappe.datetime.str_to_user(frappe.datetime.now_datetime());
	const html = `
		<!doctype html>
		<html>
		<head>
			<meta charset="utf-8">
			<title>${escape_html(__("Kautionskonten"))}</title>
			<style>
				@page { size: A4 landscape; margin: 12mm; }
				* { box-sizing: border-box; }
				body {
					margin: 0;
					color: #111827;
					font-family: Arial, sans-serif;
					font-size: 10px;
				}
				header {
					display: flex;
					justify-content: space-between;
					gap: 16px;
					margin-bottom: 10px;
					border-bottom: 1px solid #d1d5db;
					padding-bottom: 8px;
				}
				h1 {
					margin: 0 0 4px;
					font-size: 18px;
					font-weight: 700;
				}
				.meta {
					color: #4b5563;
					font-size: 9px;
					line-height: 1.35;
				}
				.summary {
					text-align: right;
					white-space: nowrap;
				}
				table {
					width: 100%;
					border-collapse: collapse;
					table-layout: fixed;
				}
				th,
				td {
					border: 1px solid #d1d5db;
					padding: 4px 5px;
					vertical-align: top;
					overflow-wrap: anywhere;
				}
				th {
					background: #f3f4f6;
					font-weight: 700;
					text-align: left;
				}
				tfoot td {
					background: #f9fafb;
					font-weight: 700;
				}
				td.number,
				th.number {
					text-align: right;
					white-space: nowrap;
				}
				.status {
					font-weight: 700;
					white-space: nowrap;
				}
				.status-ok { color: #166534; }
				.status-bad { color: #991b1b; }
				.status-warn { color: #9a3412; }
				.notes { font-size: 9px; }
				@media print {
					button { display: none; }
				}
			</style>
		</head>
		<body>
			<header>
				<div>
					<h1>${escape_html(__("Kautionskonten"))}</h1>
					<div class="meta">${filters}</div>
				</div>
				<div class="meta summary">
					${escape_html(__("Stand"))}: ${escape_html(generated_at)}<br>
					${escape_html(__("Anzahl"))}: ${rows.length}
				</div>
			</header>
			<table>
				<thead>
					<tr>${KAUTIONSKONTEN_COMPACT_COLUMNS.map(render_print_header).join("")}</tr>
				</thead>
				<tbody>
					${rows.map(render_kautionskonten_print_row).join("")}
				</tbody>
				<tfoot>
					${render_kautionskonten_print_total_row(rows)}
				</tfoot>
			</table>
			<script>
				window.onload = function () {
					window.print();
				};
			</script>
		</body>
		</html>
	`;
	print_window.document.open();
	print_window.document.write(html);
	print_window.document.close();
}

function render_print_header(column) {
	const cls = column.type === "currency" ? " class=\"number\"" : "";
	return `<th${cls}>${escape_html(column.label)}</th>`;
}

function render_kautionskonten_print_row(row) {
	return `<tr>${KAUTIONSKONTEN_COMPACT_COLUMNS.map((column) => render_print_cell(row, column)).join("")}</tr>`;
}

function render_kautionskonten_print_total_row(rows) {
	const totals = rows.reduce(
		(acc, row) => {
			acc.kaution_betrag += Number(row.kaution_betrag || 0);
			acc.saldo += Number(row.saldo || 0);
			acc.differenz += Number(row.differenz || 0);
			acc.currency = acc.currency || row.currency;
			return acc;
		},
		{ kaution_betrag: 0, saldo: 0, differenz: 0, currency: null }
	);
	const cells = KAUTIONSKONTEN_COMPACT_COLUMNS.map((column, index) => {
		if (index === 0) {
			return `<td><strong>${escape_html(__("Summe"))}</strong></td>`;
		}
		if (["kaution_betrag", "saldo", "differenz"].includes(column.fieldname)) {
			return `<td class="number"><strong>${escape_html(
				format_print_currency(totals[column.fieldname], totals.currency)
			)}</strong></td>`;
		}
		return "<td></td>";
	});
	return `<tr>${cells.join("")}</tr>`;
}

function render_print_cell(row, column) {
	let value = row[column.fieldname];
	let cls = "";
	if (column.type === "link_label" && column.fieldname === "mietvertrag") {
		value = row.mietvertrag_name || row.kunde_anzeige || row.kunde || row.mietvertrag;
	}
	if (column.type === "currency") {
		cls = "number";
		value = format_print_currency(value, row.currency);
	}
	if (column.fieldname === "pruefung") {
		cls = `status ${status_class(value)}`;
	}
	if (column.fieldname === "kaution_notizen") {
		cls = "notes";
	}
	return `<td${cls ? ` class="${cls}"` : ""}>${escape_html(value || "")}</td>`;
}

function build_kautionskonten_filter_text(report) {
	const labels = {
		company: __("Firma"),
		stichtag: __("Stichtag"),
		immobilie: __("Immobilie"),
		nur_aktive_vertraege: __("Nur aktive Mietverträge"),
		nur_mit_kautionskonto: __("Nur mit Kautionskonto"),
	};
	const parts = Object.keys(labels)
		.map((fieldname) => {
			let value = report.get_filter_value?.(fieldname);
			if (value === undefined || value === null || value === "") {
				return null;
			}
			if (fieldname === "stichtag") {
				value = frappe.datetime.str_to_user(value);
			}
			if (fieldname.startsWith("nur_")) {
				value = value ? __("Ja") : __("Nein");
			}
			return `${escape_html(labels[fieldname])}: ${escape_html(value)}`;
		})
		.filter(Boolean);
	return parts.join("<br>");
}

function format_print_currency(value, currency) {
	if (value === undefined || value === null || value === "") {
		return "";
	}
	if (frappe.format) {
		return frappe.format(value, { fieldtype: "Currency", options: "currency" }, {}, { currency });
	}
	return Number(value).toLocaleString("de-DE", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
}

function status_class(status) {
	if (status === "OK") {
		return "status-ok";
	}
	if (["Unterdeckt", "Kautionskonto fehlt"].includes(status)) {
		return "status-bad";
	}
	return "status-warn";
}

function escape_html(value) {
	return String(value)
		.replaceAll("&", "&amp;")
		.replaceAll("<", "&lt;")
		.replaceAll(">", "&gt;")
		.replaceAll('"', "&quot;")
		.replaceAll("'", "&#039;");
}
